src/components/CodeMirrorEditor.jsx

```diff
@@ -0,0 +1,90 @@
+import CodeMirror from 'codemirror';
+import isNil from 'lodash-es/isNil';
+import LRU from 'lru-cache';
+import PropTypes from 'prop-types';
+import React, {useEffect, useRef} from 'react';
+
+import 'codemirror/mode/htmlmixed/htmlmixed';
+import 'codemirror/mode/css/css';
+import 'codemirror/mode/javascript/javascript';
+
+const CODEMIRROR_MODES_MAP = {
+  html: 'htmlmixed',
+  css: 'css',
+  javascript: 'javascript',
+};
+
+export default function CodeMirrorEditor({
+  language,
+  projectKey,
+  source,
+  onInput,
+}) {
+  const containerRef = useRef();
+  const editorRef = useRef();
+  const docsCacheRef = useRef(new LRU(3));
+  const mode = CODEMIRROR_MODES_MAP[language];
+
+  useEffect(() => {
+    const {current: container} = containerRef;
+
+    const editor = editorRef.current = CodeMirror(
+      container,
+      {
+        lineNumbers: true,
+        lineWrapping: true,
+      },
+    );
```

Comment:
I know there isn't explicitly a `CodeMirror.Editor.remove()` function, but is it worth removing the DOM element on cleanup of this effect?

Reply:
It's a good thought, but I don't think so. This is a zero-dependency effect, meaning it will only be run once, after the component is initially rendered. Thus, the only time this would need to be cleaned up is when the component itself is torn down. And cleanup for a CodeMirror editor merely entails removing its container element from the DOM (see [the docs for `cm.getWrapperElement()`](https://codemirror.net/doc/manual.html#api_misc)).